// @langchain/redis/dist/schema.js
import { SchemaFieldTypes } from "redis";

//#region src/schema.ts
/**
* Default separator character for Redis TAG fields in HASH documents.
* @see https://redis.io/docs/latest/develop/ai/search-and-query/advanced-concepts/tags/
*/
const DEFAULT_TAG_SEPARATOR = ",";
/**
* Builds a RediSearch schema from metadata field definitions.
*
* This function builds up a schema based on the metadata field schema definitions.
*
* @param metadataSchema - Array of metadata field schema definitions
* @param defaultSchema - the default RediSearchSchema without considering metadata fields
* @returns a new RediSearchSchema with metadata fields added
*
* @see https://redis.io/docs/latest/develop/ai/search-and-query/indexing/schema-definition/
*
* @example
* ```typescript
* const schema: RediSearchSchema = {
*   content_vector: { type: SchemaFieldTypes.VECTOR, ... },
*   content: SchemaFieldTypes.TEXT
* };
*
* const metadataSchema: MetadataFieldSchema[] = [
*   { name: "category", type: "tag" },
*   { name: "price", type: "numeric", options: { sortable: true } }
* ];
*
* const updatedSchema = buildMetadataSchema(metadataSchema, schema);
* // updatedSchema includes category and price fields
* ```
*/
function buildMetadataSchema(metadataSchema, defaultSchema) {
	const updatedSchema = { ...defaultSchema };
	for (const fieldSchema of metadataSchema) switch (fieldSchema.type) {
		case "tag": {
			const tagOptions = {
				type: SchemaFieldTypes.TAG,
				SEPARATOR: fieldSchema.options?.separator || DEFAULT_TAG_SEPARATOR
			};
			if (fieldSchema.options?.caseSensitive) tagOptions.CASESENSITIVE = true;
			if (fieldSchema.options?.noindex) tagOptions.NOINDEX = true;
			updatedSchema[fieldSchema.name] = tagOptions;
			break;
		}
		case "text": {
			const textOptions = { type: SchemaFieldTypes.TEXT };
			if (fieldSchema.options?.weight !== void 0) textOptions.WEIGHT = fieldSchema.options.weight;
			if (fieldSchema.options?.noStem) textOptions.NOSTEM = true;
			if (fieldSchema.options?.noindex) textOptions.NOINDEX = true;
			if (fieldSchema.options?.sortable) textOptions.SORTABLE = true;
			updatedSchema[fieldSchema.name] = textOptions;
			break;
		}
		case "numeric": {
			const numericOptions = { type: SchemaFieldTypes.NUMERIC };
			if (fieldSchema.options?.sortable) numericOptions.SORTABLE = true;
			if (fieldSchema.options?.noindex) numericOptions.NOINDEX = true;
			updatedSchema[fieldSchema.name] = numericOptions;
			break;
		}
		case "geo": {
			const geoOptions = { type: SchemaFieldTypes.GEO };
			if (fieldSchema.options?.noindex) geoOptions.NOINDEX = true;
			updatedSchema[fieldSchema.name] = geoOptions;
			break;
		}
		default: updatedSchema[fieldSchema.name] = { type: SchemaFieldTypes.TEXT };
	}
	return updatedSchema;
}
/**
* Serializes metadata field values for storage in Redis based on field type.
*
* Converts JavaScript values to the appropriate format for Redis storage:
* - Tag fields: Arrays joined with separator, or string values
* - Text fields: String values
* - Numeric fields: Number values (Date objects are automatically converted to Unix epoch timestamps in seconds)
* - Geo fields: "longitude,latitude" string format
* - Geoshape fields: WKT (Well-Known Text) format strings
*
* @param fieldSchema - The metadata field schema definition
* @param fieldValue - The value to serialize
* @returns The serialized value ready for Redis storage
*
* @see https://redis.io/docs/latest/develop/ai/search-and-query/indexing/field-and-type-options/
*
* @example
* ```typescript
* const tagSchema = { name: "category", type: "tag" as const };
* serializeMetadataField(tagSchema, ["electronics", "gadgets"]);
* // Returns: "electronics|gadgets"
*
* const geoSchema = { name: "location", type: "geo" as const };
* serializeMetadataField(geoSchema, [-122.4194, 37.7749]);
* // Returns: "-122.4194,37.7749"
*
* const geoshapeSchema = { name: "area", type: "geoshape" as const };
* serializeMetadataField(geoshapeSchema, "POLYGON((0 0, 0 10, 10 10, 10 0, 0 0))");
* // Returns: "POLYGON((0 0, 0 10, 10 10, 10 0, 0 0))"
*
* const numericSchema = { name: "created_at", type: "numeric" as const };
* serializeMetadataField(numericSchema, new Date("2023-01-01"));
* // Returns: 1672531200 (Unix epoch timestamp in seconds)
*
* serializeMetadataField(numericSchema, 42);
* // Returns: 42
* ```
*/
function serializeMetadataField(fieldSchema, fieldValue) {
	switch (fieldSchema.type) {
		case "tag": return Array.isArray(fieldValue) ? fieldValue.join(fieldSchema.options?.separator || DEFAULT_TAG_SEPARATOR) : String(fieldValue);
		case "text": return String(fieldValue);
		case "numeric":
			if (fieldValue && typeof fieldValue === "object" && "getTime" in fieldValue && typeof fieldValue.getTime === "function") return Math.floor(fieldValue.getTime() / 1e3);
			return Number(fieldValue);
		case "geo":
			if (Array.isArray(fieldValue) && fieldValue.length === 2) return `${fieldValue[0]},${fieldValue[1]}`;
			return String(fieldValue);
		default: return String(fieldValue);
	}
}
/**
* Deserializes metadata field values from Redis storage based on field type.
*
* Converts Redis-stored values back to JavaScript types:
* - Tag fields: Splits separator-delimited strings back to arrays (if separator is present)
* - Numeric fields: Converts to numbers (keeps as number, does NOT convert to Date)
* - Geo fields: Converts "longitude,latitude" strings to [lon, lat] arrays
* - Text fields: Returns as-is
*
* Note: Numeric fields are returned as numbers. If you stored a Date object as a Unix epoch
* timestamp, you'll need to manually convert it back to a Date object if needed:
* `new Date(numericValue * 1000)`
*
* @param fieldSchema - The metadata field schema definition
* @param fieldValue - The value from Redis to deserialize
* @returns The deserialized value in JavaScript format
*
* @see https://redis.io/docs/latest/develop/ai/search-and-query/indexing/field-and-type-options/
*
* @example
* ```typescript
* const tagSchema = { name: "category", type: "tag" as const };
* deserializeMetadataField(tagSchema, "electronics|gadgets");
* // Returns: ["electronics", "gadgets"]
*
* const geoSchema = { name: "location", type: "geo" as const };
* deserializeMetadataField(geoSchema, "-122.4194,37.7749");
* // Returns: [-122.4194, 37.7749]
*
* const geoshapeSchema = { name: "area", type: "geoshape" as const };
* deserializeMetadataField(geoshapeSchema, "POLYGON((0 0, 0 10, 10 10, 10 0, 0 0))");
* // Returns: "POLYGON((0 0, 0 10, 10 10, 10 0, 0 0))"
*
* const numericSchema = { name: "created_at", type: "numeric" as const };
* deserializeMetadataField(numericSchema, "1672531200");
* // Returns: 1672531200 (number)
* // To convert to Date: new Date(1672531200 * 1000)
*
* deserializeMetadataField(numericSchema, "42");
* // Returns: 42
* ```
*/
function deserializeMetadataField(fieldSchema, fieldValue) {
	if (fieldValue === void 0 || fieldValue === null) return fieldValue;
	switch (fieldSchema.type) {
		case "tag": {
			const separator = fieldSchema.options?.separator || DEFAULT_TAG_SEPARATOR;
			if (typeof fieldValue === "string" && fieldValue.includes(separator)) return fieldValue.split(separator);
			return fieldValue;
		}
		case "numeric": return Number(fieldValue);
		case "geo":
			if (typeof fieldValue === "string" && fieldValue.includes(",")) {
				const [lon, lat] = fieldValue.split(",").map(Number);
				return [lon, lat];
			}
			return fieldValue;
		default: return fieldValue;
	}
}
/**
* Infers metadata schema from a collection of documents by analyzing their metadata fields.
*
* This function examines the metadata of all provided documents and attempts to infer
* the appropriate field type for each metadata key based on the values found.
*
* Type inference rules:
* - Strings in "lon,lat" format → geo
* - Numbers or Date objects → numeric
* - Array of any type → tag
* - All other types → text
*
* @param documents - Array of documents to analyze
* @returns Array of inferred metadata field schemas
*
* @example
* ```typescript
* const documents = [
*   { pageContent: "...", metadata: { category: "electronics", price: 99 } },
*   { pageContent: "...", metadata: { category: "books", price: 15 } },
* ];
*
* const schema = inferMetadataSchema(documents);
* // Returns: [
* //   { name: "text", type: "tag" },
* //   { name: "price", type: "numeric" }
* // ]
* ```
*/
function inferMetadataSchema(documents) {
	if (!documents || documents.length === 0) return [];
	const fieldValues = /* @__PURE__ */ new Map();
	for (const doc of documents) {
		if (!doc.metadata || typeof doc.metadata !== "object") continue;
		for (const [key, value] of Object.entries(doc.metadata)) {
			if (value === void 0 || value === null) continue;
			if (!fieldValues.has(key)) fieldValues.set(key, []);
			fieldValues.get(key).push(value);
		}
	}
	const schema = [];
	for (const [fieldName, values] of fieldValues.entries()) {
		const fieldType = inferFieldType(values);
		schema.push({
			name: fieldName,
			type: fieldType
		});
	}
	return schema;
}
/**
* Checks if two metadata schemas have a mismatch.
*
* This function compares two metadata schema arrays to determine if they contain
* the same fields with matching types. The comparison is order-independent and
* only considers the non-optional properties (name and type) of each field.
*
* @param customSchema - The custom metadata schema to compare
* @param inferredSchema - The inferred metadata schema to compare against
* @returns `true` if there is a mismatch, `false` if the schemas match
*
* @example
* ```typescript
* const customSchema = [
*   { name: "category", type: "tag" },
*   { name: "price", type: "numeric" }
* ];
*
* const inferredSchema = [
*   { name: "price", type: "numeric" },
*   { name: "category", type: "tag" }
* ];
*
* checkForSchemaMismatch(customSchema, inferredSchema);
* // Returns: false (schemas match, order doesn't matter)
*
* const mismatchedSchema = [
*   { name: "category", type: "text" }, // Different type
*   { name: "price", type: "numeric" }
* ];
*
* checkForSchemaMismatch(customSchema, mismatchedSchema);
* // Returns: true (type mismatch for "category")
* ```
*/
function checkForSchemaMismatch(customSchema, inferredSchema) {
	if (customSchema.length !== inferredSchema.length) return true;
	const inferredMap = /* @__PURE__ */ new Map();
	for (const field of inferredSchema) inferredMap.set(field.name, field);
	for (const customField of customSchema) {
		const inferredField = inferredMap.get(customField.name);
		if (!inferredField) return true;
		if (customField.type !== inferredField.type) return true;
	}
	return false;
}
/**
* Infers the appropriate field type for a metadata field based on its values.
*
* @param values - Array of sample values for this field
* @returns The inferred field type
*/
function inferFieldType(values) {
	if (values.length === 0) return "text";
	const allGeo = values.every((value) => isGeoCoordinate(value));
	if (allGeo) return "geo";
	const allNumeric = values.every((value) => isNumberOrDate(value));
	if (allNumeric) return "numeric";
	const allArrays = values.every((value) => Array.isArray(value));
	if (allArrays) return "tag";
	return "text";
}
/**
* Checks if a value represents a geo coordinate.
*
* @param value - The value to check
* @returns True if the value is a geo coordinate
*/
function isGeoCoordinate(value) {
	if (typeof value === "string") {
		const parts = value.split(",");
		if (parts.length === 2) {
			const lon = parseFloat(parts[0].trim());
			const lat = parseFloat(parts[1].trim());
			return !Number.isNaN(lon) && !Number.isNaN(lat);
		}
	}
	return false;
}
/**
* Checks if a value is a number or a Date object.
*
* @param value - The value to check
* @returns True if the value is a number or Date
*/
function isNumberOrDate(value) {
	return typeof value === "number" || typeof value === "object" && value !== null && "getTime" in value && typeof value.getTime === "function";
}
/**
* Converts legacy CustomSchemaField format to new MetadataFieldSchema format.
*
* This function provides backward compatibility by converting the old Record-based
* schema format to the new array-based format. It also emits a deprecation warning.
*
* @param metadataKey the custom metadata key prefix
* @param legacySchema - The legacy schema in Record<string, CustomSchemaField> format
* @returns The converted schema in MetadataFieldSchema[] format
*
* @example
* ```typescript
* const legacySchema = {
*   userId: { type: SchemaFieldTypes.TAG, SEPARATOR: "|" },
*   price: { type: SchemaFieldTypes.NUMERIC, SORTABLE: true }
* };
*
* const newSchema = convertLegacySchema(legacySchema);
* // Returns: [
* //   { name: "userId", type: "tag", options: { separator: "|" } },
* //   { name: "price", type: "numeric", options: { sortable: true } }
* // ]
* ```
*/
function convertLegacySchema(metadataKey, legacySchema) {
	console.warn("DEPRECATION WARNING: The Record<string, CustomSchemaField> format for customSchema is deprecated. Please migrate to the new MetadataFieldSchema[] format. See https://js.langchain.com/docs/integrations/vectorstores/redis for migration guide. This legacy format will be removed in the next major version.");
	const convertedSchema = [];
	for (const [fieldName, fieldConfig] of Object.entries(legacySchema)) {
		let type;
		switch (fieldConfig.type) {
			case SchemaFieldTypes.TAG:
				type = "tag";
				break;
			case SchemaFieldTypes.TEXT:
				type = "text";
				break;
			case SchemaFieldTypes.NUMERIC:
				type = "numeric";
				break;
			case SchemaFieldTypes.GEO:
				type = "geo";
				break;
			default:
				type = "text";
				console.warn(`Unknown field type ${fieldConfig.type} for field ${fieldName}, defaulting to "text"`);
		}
		const options = {};
		if (fieldConfig.SEPARATOR !== void 0) options.separator = fieldConfig.SEPARATOR;
		if (fieldConfig.CASESENSITIVE !== void 0) options.caseSensitive = fieldConfig.CASESENSITIVE;
		if (fieldConfig.WEIGHT !== void 0) options.weight = fieldConfig.WEIGHT;
		if (fieldConfig.NOSTEM !== void 0) options.noStem = fieldConfig.NOSTEM;
		if (fieldConfig.SORTABLE !== void 0) options.sortable = fieldConfig.SORTABLE === true || fieldConfig.SORTABLE === "UNF";
		if (fieldConfig.NOINDEX !== void 0) options.noindex = fieldConfig.NOINDEX;
		convertedSchema.push({
			name: `${metadataKey}.${fieldName}`,
			type,
			...Object.keys(options).length > 0 ? { options } : {}
		});
	}
	return convertedSchema;
}

//#endregion
export { DEFAULT_TAG_SEPARATOR, buildMetadataSchema, checkForSchemaMismatch, convertLegacySchema, deserializeMetadataField, inferMetadataSchema, serializeMetadataField };
