// @langchain/redis/dist/vectorstores_fluent.cjs
const require_rolldown_runtime = require('./_virtual/rolldown_runtime.cjs');
const require_filters = require('./filters.cjs');
const require_schema = require('./schema.cjs');
const redis = require_rolldown_runtime.__toESM(require("redis"));
const __langchain_core_documents = require_rolldown_runtime.__toESM(require("@langchain/core/documents"));
const __langchain_core_vectorstores = require_rolldown_runtime.__toESM(require("@langchain/core/vectorstores"));
const uuid = require_rolldown_runtime.__toESM(require("uuid"));

//#region src/vectorstores_fluent.ts
/**
* Advanced Redis Vector Store with structured metadata filtering.
*
* This class provides advanced filtering capabilities through FilterExpression
* and requires explicit MetadataFieldSchema definition. It supports:
* - Tag filters for categorical data
* - Numeric filters for ranges and comparisons
* - Text filters for full-text search
* - Geo filters for location-based queries
* - Complex filter combinations with AND/OR logic
*
* For basic filtering with simple string[] or string filters, use RedisVectorStore instead.
*
* @example
* ```typescript
* const vectorStore = await FluentRedisVectorStore.fromDocuments(
*   docs,
*   embeddings,
*   {
*     redisClient: client,
*     indexName: "products",
*     customSchema: [
*       { name: "category", type: "tag" },
*       { name: "price", type: "numeric", options: { sortable: true } }
*     ]
*   }
* );
*
* // Use advanced filtering
* const results = await vectorStore.similaritySearch(
*   "laptop",
*   5,
*   Tag("category").eq("electronics").and(Num("price").lt(1000))
* );
* ```
*/
var FluentRedisVectorStore = class FluentRedisVectorStore extends __langchain_core_vectorstores.VectorStore {
	redisClient;
	indexName;
	indexOptions;
	createIndexOptions;
	keyPrefix;
	contentKey;
	metadataKey;
	vectorKey;
	filter;
	ttl;
	customSchema;
	_vectorstoreType() {
		return "redis_fluent";
	}
	constructor(embeddings, _dbConfig) {
		super(embeddings, _dbConfig);
		this.redisClient = _dbConfig.redisClient;
		this.indexName = _dbConfig.indexName;
		this.indexOptions = _dbConfig.indexOptions ?? {
			ALGORITHM: redis.VectorAlgorithms.HNSW,
			DISTANCE_METRIC: "COSINE"
		};
		this.keyPrefix = _dbConfig.keyPrefix ?? `doc:${this.indexName}:`;
		this.contentKey = _dbConfig.contentKey ?? "content";
		this.metadataKey = _dbConfig.metadataKey ?? "metadata";
		this.vectorKey = _dbConfig.vectorKey ?? "content_vector";
		this.filter = _dbConfig.filter;
		this.ttl = _dbConfig.ttl;
		this.customSchema = _dbConfig.customSchema;
		this.createIndexOptions = {
			ON: "HASH",
			PREFIX: this.keyPrefix,
			..._dbConfig.createIndexOptions
		};
	}
	/**
	* Method for adding documents to the RedisVectorStore. It first converts
	* the documents to texts and then adds them as vectors.
	* @param documents The documents to add.
	* @param options Optional parameters for adding the documents.
	* @returns A promise that resolves when the documents have been added.
	*/
	async addDocuments(documents, options) {
		const texts = documents.map(({ pageContent }) => pageContent);
		return this.addVectors(await this.embeddings.embedDocuments(texts), documents, options);
	}
	/**
	* Method for adding vectors to the FluentRedisVectorStore. It checks if the
	* index exists and creates it if it doesn't, then adds the vectors in batches.
	* @param vectors The vectors to add.
	* @param documents The documents associated with the vectors.
	* @param keys Optional keys for the vectors.
	* @param batchSize The size of the batches in which to add the vectors. Defaults to 1000.
	* @returns A promise that resolves when the vectors have been added.
	*/
	async addVectors(vectors, documents, { keys, batchSize = 1e3 } = {}) {
		if (!vectors.length || !vectors[0].length) throw new Error("No vectors provided");
		await this.createIndex(documents, vectors[0].length);
		const multi = this.redisClient.multi();
		await Promise.all(vectors.map(async (vector, idx) => {
			const key = keys && keys.length ? keys[idx] : `${this.keyPrefix}${(0, uuid.v4)()}`;
			const metadata = documents[idx] && documents[idx].metadata ? documents[idx].metadata : {};
			const hashFields = {
				[this.vectorKey]: this.getFloat32Buffer(vector),
				[this.contentKey]: documents[idx].pageContent
			};
			if (this.customSchema && this.customSchema.length > 0) for (const fieldSchema of this.customSchema) {
				const fieldValue = metadata[fieldSchema.name];
				if (fieldValue !== void 0 && fieldValue !== null) hashFields[fieldSchema.name] = require_schema.serializeMetadataField(fieldSchema, fieldValue);
			}
			multi.hSet(key, hashFields);
			if (this.ttl) multi.expire(key, this.ttl);
			if (idx % batchSize === 0) await multi.exec();
		}));
		await multi.exec();
	}
	/**
	* Method for performing a similarity search in the FluentRedisVectorStore.
	* Returns documents and their similarity scores.
	* @param query The query vector.
	* @param k The number of nearest neighbors to return.
	* @param filter Optional FilterExpression to apply to the search.
	* @returns A promise that resolves to an array of documents and their scores.
	*/
	async similaritySearchVectorWithScore(query, k, filter) {
		if (filter && this.filter) throw new Error("cannot provide both `filter` and `this.filter`");
		const _filter = filter ?? this.filter;
		const results = await this.redisClient.ft.search(this.indexName, ...this.buildQuery(query, k, _filter));
		const result = [];
		if (results.total) {
			for (const res of results.documents) if (res.value) {
				const document = res.value;
				if (document.vector_score) {
					const metadata = {};
					if (this.customSchema && this.customSchema.length > 0) for (const fieldSchema of this.customSchema) {
						const fieldValue = document[fieldSchema.name];
						if (fieldValue !== void 0 && fieldValue !== null) metadata[fieldSchema.name] = require_schema.deserializeMetadataField(fieldSchema, fieldValue);
					}
					result.push([new __langchain_core_documents.Document({
						pageContent: document[this.contentKey] ?? "",
						metadata
					}), Number(document.vector_score)]);
				}
			}
		}
		return result;
	}
	/**
	* Static method for creating a new instance of FluentRedisVectorStore from
	* texts. It creates documents from the texts and metadata, then adds them
	* to the FluentRedisVectorStore.
	* @param texts The texts to add.
	* @param metadatas The metadata associated with the texts.
	* @param embeddings The embeddings to use.
	* @param dbConfig The configuration for the FluentRedisVectorStore.
	* @param docsOptions The document options to use.
	* @returns A promise that resolves to a new instance of FluentRedisVectorStore.
	*/
	static fromTexts(texts, metadatas, embeddings, dbConfig, docsOptions) {
		const docs = [];
		for (let i = 0; i < texts.length; i += 1) {
			const metadata = Array.isArray(metadatas) ? metadatas[i] : metadatas;
			const newDoc = new __langchain_core_documents.Document({
				pageContent: texts[i],
				metadata
			});
			docs.push(newDoc);
		}
		return FluentRedisVectorStore.fromDocuments(docs, embeddings, dbConfig, docsOptions);
	}
	/**
	* Static method for creating a new instance of FluentRedisVectorStore from
	* documents. It adds the documents to the FluentRedisVectorStore.
	* @param docs The documents to add.
	* @param embeddings The embeddings to use.
	* @param dbConfig The configuration for the FluentRedisVectorStore.
	* @param docsOptions The document options to use.
	* @returns A promise that resolves to a new instance of FluentRedisVectorStore.
	*/
	static async fromDocuments(docs, embeddings, dbConfig, docsOptions) {
		const instance = new this(embeddings, dbConfig);
		await instance.addDocuments(docs, docsOptions);
		return instance;
	}
	/**
	* Method for checking if an index exists in the RedisVectorStore.
	* @returns A promise that resolves to a boolean indicating whether the index exists.
	*/
	async checkIndexState() {
		try {
			const result = await this.redisClient.ft.info(this.indexName);
			return result.attributes.some((attr) => attr.identifier === this.metadataKey) ? "legacy" : "default";
		} catch (err) {
			if (err?.message.includes("unknown command")) throw new Error("Failed to run FT.INFO command. Please ensure that you are running a RediSearch-capable Redis instance: https://js.langchain.com/docs/integrations/vectorstores/redis/#setup");
			return "none";
		}
	}
	/**
	* Method for creating an index in the FluentRedisVectorStore.
	* Requires a customSchema to be defined for proper metadata indexing.
	* If the index already exists, it does nothing.
	*
	* @param documents Optional documents to validate against the schema
	* @param dimensions The dimensions of the vector field (default: 1536)
	* @returns A promise that resolves when the index has been created.
	* @throws Error if customSchema is not defined
	*/
	async createIndex(documents, dimensions = 1536) {
		if (!this.customSchema || this.customSchema.length === 0) throw new Error("FluentRedisVectorStore requires a customSchema to be defined. Please provide a customSchema with MetadataFieldSchema definitions in the configuration.");
		const indexState = await this.checkIndexState();
		if (documents && documents.length > 0) {
			const inferredSchema = require_schema.inferMetadataSchema(documents);
			if (require_schema.checkForSchemaMismatch(this.customSchema, inferredSchema)) console.warn("The custom schema does not match the metadata schema inferred from the documents. This is not necessarily an issue, but could indicate an invalid custom schema.");
		}
		let schema = {
			[this.vectorKey]: {
				type: redis.SchemaFieldTypes.VECTOR,
				TYPE: "FLOAT32",
				DIM: dimensions,
				...this.indexOptions
			},
			[this.contentKey]: redis.SchemaFieldTypes.TEXT
		};
		schema = require_schema.buildMetadataSchema(this.customSchema, schema);
		if (indexState === "none") await this.redisClient.ft.create(this.indexName, schema, this.createIndexOptions);
	}
	/**
	* Method for dropping an index from the RedisVectorStore.
	* @param deleteDocuments Optional boolean indicating whether to drop the associated documents.
	* @returns A promise that resolves to a boolean indicating whether the index was dropped.
	*/
	async dropIndex(deleteDocuments) {
		try {
			const options = deleteDocuments ? { DD: deleteDocuments } : void 0;
			await this.redisClient.ft.dropIndex(this.indexName, options);
			return true;
		} catch {
			return false;
		}
	}
	/**
	* Deletes vectors from the vector store.
	*
	* Supports two deletion modes:
	* - Delete all documents by dropping the entire index and recreating it
	* - Delete specific documents by their IDs using Redis DEL operation
	*
	* @param params - The deletion parameters. Must be one of:
	*   - `{ deleteAll: boolean }` - If true, drops the entire index and all associated documents
	*   - `{ ids: string[] }` - Array of document IDs to delete. IDs will be automatically prefixed with the configured keyPrefix
	* @returns A promise that resolves when the deletion operation is complete
	* @throws {Error} Throws an error if invalid parameters are provided (neither deleteAll nor ids specified)
	*
	* @example
	* Delete all documents:
	* ```typescript
	* await vectorStore.delete({ deleteAll: true });
	* ```
	*
	* @example
	* Delete specific documents by ID:
	* ```typescript
	* await vectorStore.delete({ ids: ['doc1', 'doc2', 'doc3'] });
	* ```
	*/
	async delete(params) {
		if ("deleteAll" in params && params.deleteAll) await this.dropIndex(true);
		else if ("ids" in params && params.ids && params.ids.length > 0) {
			const keys = params.ids.map((id) => `${this.keyPrefix}${id}`);
			await this.redisClient.del(keys);
		} else throw new Error(`Invalid parameters passed to "delete".`);
	}
	buildQuery(query, k, filter) {
		const vectorScoreField = "vector_score";
		let hybridFields = "*";
		if (filter) hybridFields = this.prepareFilter(filter);
		const baseQuery = `${hybridFields} => [KNN ${k} @${this.vectorKey} $vector AS ${vectorScoreField}]`;
		const returnFields = [this.contentKey, vectorScoreField];
		if (this.customSchema) for (const fieldName of this.customSchema) returnFields.push(`${fieldName.name}`);
		const options = {
			PARAMS: { vector: this.getFloat32Buffer(query) },
			RETURN: returnFields,
			SORTBY: vectorScoreField,
			DIALECT: 2,
			LIMIT: {
				from: 0,
				size: k
			}
		};
		return [baseQuery, options];
	}
	prepareFilter(filter) {
		if (!filter) return "*";
		if (typeof filter === "object" && "toString" in filter && typeof filter.toString === "function") return filter.toString();
		throw new Error("FluentRedisVectorStore only supports FilterExpression filters. Use Tag(), Num(), Text(), Geo(), or Timestamp() to create filters. For simple string[] or string filters, use the basic RedisVectorStore instead.");
	}
	/**
	* Converts the vector to the buffer Redis needs to
	* correctly store an embedding
	*
	* @param vector
	* @returns Buffer
	*/
	getFloat32Buffer(vector) {
		return Buffer.from(new Float32Array(vector).buffer);
	}
};
/**
* @deprecated Use {@link FluentRedisVectorStore} instead.
* This is an alias for backward compatibility.
*/
const RedisVectorStoreAdvanced = FluentRedisVectorStore;

//#endregion
exports.FluentRedisVectorStore = FluentRedisVectorStore;
exports.RedisVectorStoreAdvanced = RedisVectorStoreAdvanced;
